ifsbench/benchmark.py

```diff
@@ -0,0 +1,194 @@
+# (C) Copyright 2020- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+"""
+Generic benchmark implementation.
+"""
+
+from dataclasses import dataclass, field
+import os
+from pathlib import Path
+from typing import List, Optional
+
+from ifsbench.application import Application
+from ifsbench.arch import Arch
+from ifsbench.data import DataHandler
+from ifsbench.env import EnvHandler, DefaultEnvPipeline
+from ifsbench.job import Job
+from ifsbench.launch import Launcher
+
+
+__all__ = ['ScienceSetup', 'TechSetup', 'Benchmark']
+
+@dataclass
+class ScienceSetup:
+    """
+    Generic (scientific) benchmark setup.
+
+    This dataclass encapsulates the information that is needed to describe a 
+    benchmark:
+      * The application that is benchmarked.
```

Comment:
Can you add something here along the lines of 'unless an application in TechSetup is provided' or '; will be overridden by TechSetup.application if provided'?